### PR TITLE
[CompilerPerf] Reduce memory footprint for Typar type

### PR DESCRIPTION
**Pull request by AviAvni**
@dsyme I did the same thing as for Entity and Val and the build failed and I can't find why
please help

### CODE REVIEW

File: src/fsharp/tast.fs [hunk 1]
```diff
@@ -2042,7 +2065,10 @@ and
         | _ -> ty
 
     /// Indicates if a type variable has been linked. Only used during unpickling of F# metadata.
-    member x.IsLinked = match box x.typar_attribs with null -> false | _ -> true 
+    member x.IsLinked = 
+        match x.typar_stamp with 
+        | 0L -> false
```
Comment (dsyme) [inline src/fsharp/tast.fs]: This looks a tiny bit dubious - 0 is a valid stamp, but perhaps it's never used

**Comment by dsyme** [comment]
@AviAvni I edited your PR to try to see if stamp 0 was the problem

**Comment by dsyme** [comment]
@AviAvni It looks like the same Typar can have `.Link` called multiple times.  This is somewhat unexpected, but I've adjusted the code to do what it used to do logically speaking in this case

**Comment by dsyme** [comment]
@AviAvni Ugh, still doesn't work

**Comment by dsyme** [comment]
OK, copyTypar needed to clone the mutable data, see commits above. Ugly!

**Comment by smoothdeveloper** [comment]
Can we consider a bit of formatting in `copyTypar`? It will make the ugly code a bit easier to read.

```fsharp
let copyTypar (tp: Typar) = 
    let optData =
        // Be careful to clone the mutable optional data too
        match tp.typar_opt_data with
        | None -> None
        | Some tg -> 
            Some
                ({ 
                    typar_il_name = tg.typar_il_name
                    typar_xmldoc = tg.typar_xmldoc
                    typar_constraints = tg.typar_constraints 
                })
                
    Typar.New 
        { tp with 
            typar_stamp = newStamp()
            typar_astype = Unchecked.defaultof<_>
            typar_opt_data = optData
        } 
```

(notepad edit)

**Comment by AviAvni** [comment]
@dsyme thank you very much I'll continue the work and post the statistics soon
@smoothdeveloper thanks I'll beautify copyTypar

**Comment by dsyme** [comment]
> @smoothdeveloper thanks I'll beautify copyTypar

I think just use an explicit full copy (no copy-and-update `{ x with y ..}`)

**Comment by AviAvni** [comment]
@dsyme changed to full copy

**Comment by AviAvni** [comment]
@dsyme here are the statistics for fsharp.compiler.fsproj
the code for the measurment https://github.com/Microsoft/visualfsharp/compare/master...AviAvni:fcs-memory-3-measurment
```
~~~ Typar statistics: WithOptData 479000, WithoutOptData 254858
~~~ Typar statistics: typar_il_name: 0
~~~ Typar statistics: typar_xmldoc: 0
~~~ Typar statistics: typar_constraints: 479000
~~~ Typar statistics: typar_attribs: 479000
```

**Comment by AviAvni** [comment]
@dotnet-bot test this please

**Comment by dsyme** [comment]
@dotnet-bot test this please

**Comment by AviAvni** [comment]
@dsyme  updated statistics
```
~~~ Typar statistics: WithOptData 66920, WithoutOptData 666474
~~~ Typar statistics: typar_il_name: 0
~~~ Typar statistics: typar_xmldoc: 0
~~~ Typar statistics: typar_constraints: 66456
~~~ Typar statistics: typar_attribs: 1190
```

**Comment by dsyme** [comment]
@AviAvni great!

**Comment by dsyme** [comment]
@dotnet-bot test this please

**Comment by dsyme** [comment]
@dotnet-bot test this please

**Comment by dsyme** [comment]
@AviAvni Tests triggered now :)

**Comment by AviAvni** [comment]
@dsyme can this be merged or there is something to do?

**Comment by realvictorprm** [comment]
@AviAvni i guess being at Fsharpx prevents us :p

**Comment by KevinRansom** [comment]
@dotnet-bot test this please

**Comment by cartermp** [comment]
Merging as per approvals.

**Comment by cartermp** [comment]
Thanks for this awesome work (as always 😄), @AviAvni

**Pull request merged**